packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1048,6 +1087,11 @@ function useMutableSource<Source, Snapshot>(
       const maybeNewSnapshot = getSnapshot(source._source);
       if (!is(snapshot, maybeNewSnapshot)) {
         setSnapshot(maybeNewSnapshot);
+
+        // We missed a mutation before committing.
+        // It's possible that other components using this source also have pending updates scheduled.
+        // In that case, we should ensure they all commit together.
+        markRootExpiredAtTime(root, getLastPendingExpirationTime(root));
```

Comment:
Need to mark this new `setSnapshot` call as a pending mutation, too, right before marking it as expired.

Reply:
This change makes sense conceptually, but it breaks the [most recent test case](https://github.com/facebook/react/pull/18297#issuecomment-606898157) I added.

In this new test I...
1. Mount and flush component A with data version 1.
2. Mount component B with data version 1 (but flush passive effects).
3. Mutate the data (which schedules an update for already mounted component A.
4. Flush passive effects (which also schedules an update for the newly mounted component B).
5. Verify there was no tearing.

**Before** this change, both components A and B rendered together after we flushed passive effects.

**After** this change, component B renders and commits, then component A (and we tear in between). This is because- with the change in place, we're marking the root as expired with a higher priority (e.g. 1073741296 instead of 2/idle).

I think this indicates that we should be marking root as expired with the *first* expiration time rather than the *last* in this case? At least, doing that fixes this test (and also the other two failing tests that were tearing before). This also means that we aren't *directly* using last expiration time for anything at this point (although indicate it determines when we reset first expiration time in `clearPendingUpdates`).

Reply:
I'm going to push this change as a separate commit (rather than squashing) just to give you the opportunity to review it more closely: 7c9181f

Back to you @acdlite